syntax = "proto3";

option java_multiple_files = true;
option java_package = "cn.abellee.cniface.grpc";
option java_outer_classname = "CNIFaceProto";
option objc_class_prefix = "CNIFACE";

package cniface;

import "google/protobuf/empty.proto";


service DetectService {
    rpc detect (DetectRequest) returns ( DetectResponse) {}
}

message DetectRequest {
    string model = 1;
    string faceImageBase64 = 2;
    float score = 3;
    bool isReturnRecognitionSubImage = 4;
}

message DetectResult {
    float score = 1;
    float x = 2;
    float y = 3;
    float w = 4;
    float h = 5;
    repeated float kps = 6;
    float maskScore = 7;
    string subImageBase64 = 8;
}

message DetectResponse {
    int32 code = 1;
    string message = 2;
    repeated DetectResult results = 3;
}


service RecognitionService {
    rpc extractFeature (ExtractFeatureRequest) returns ( ExtractFeatureResponse) {}
    rpc similarity (SimilarityRequest) returns ( SimilarityResponse) {}
}

message ExtractFeatureRequest {
    string model = 1;
    string faceImageBase64 = 2;
    repeated float kps = 3;
    bool isDoNormalizeL2 = 4;
}

message ExtractFeatureResponse {
    int32 code = 1;
    string message = 2;
    repeated float feature = 3;
}

message SimilarityRequest {
    repeated float feature1 = 1;
    repeated float feature2 = 2;
}

message SimilarityResponse {
    int32 code = 1;
    string message = 2;
    float similarity = 3;
}


service AttributeService {
    rpc genderAge (GenderAgeRequest) returns ( GenderAgeResponse) {}
}

message GenderAgeRequest {
    string model = 1;
    string faceImageBase64 = 2;
    repeated float kps = 3;
}

message GenderAgeResponse {
    int32 code = 1;
    string message = 2;
    int32 gender = 3;
    int32 age = 4;
}


// RepositoryService
message CreateRepositoryRequest {
    string repoName = 1;
}

message CreateRepositoryResponse {
    int32 code = 1;
    string message = 2;
}

message DeleteRepositoryRequest {
    string repoName = 1;
}

message DeleteRepositoryResponse {
    int32 code = 1;
    string message = 2;
}

message ExistRepositoryRequest {
    string repoName = 1;
}

message ExistRepositoryResponse {
    int32 code = 1;
    string message = 2;
    bool exist = 3;
}

message ListRepositoryResponse {
    int32 code = 1;
    string message = 2;
    repeated string names = 3;
}

message AddOrUpdateItemRequest {
    string repoName = 1;
    int64 id = 2;
    repeated float feature = 3;
}

message AddOrUpdateItemResponse {
    int32 code = 1;
    string message = 2;
}

message DeleteItemRequest {
    string repoName = 1;
    int64 id = 2;
}

message DeleteItemResponse {
    int32 code = 1;
    string message = 2;
}

message SearchResponseItem {
    int64 id = 1;
    float score = 2;
}

message SearchRequest {
    string repoName = 1;
    int32 topk = 2;
    repeated float feature = 3;
}

message SearchResponse {
    int32 code = 1;
    string message = 2;
    repeated SearchResponseItem results = 3;
}

service RepositoryService {
    rpc createRepository (CreateRepositoryRequest) returns ( CreateRepositoryResponse) {}
    rpc deleteRepository (DeleteRepositoryRequest) returns ( DeleteRepositoryResponse) {}
    rpc existRepository (ExistRepositoryRequest) returns ( ExistRepositoryResponse) {}
    rpc listRepository (google.protobuf.Empty) returns ( ListRepositoryResponse) {}
    rpc addOrUpdateItem (AddOrUpdateItemRequest) returns ( AddOrUpdateItemResponse) {}
    rpc deleteItem (DeleteItemRequest) returns ( DeleteItemResponse) {}
    rpc search (SearchRequest) returns ( SearchResponse) {}
}


message OCRRequest {
    string imageBase64 = 1;
    bool isCls = 2; // 设置为true，可识别倒立的文字。
    bool isRec = 3; // 是否识别文字。
}

message OCRResponseItem {
    repeated float position = 1;
    float score = 2;
    string text = 3;
    float clsScore = 4;
    int32 clsLabel = 5;
}

message OCRResponse {
    int32 code = 1;
    string message = 2;
    repeated OCRResponseItem results = 3;
}

service OCRService {
    rpc ocr (OCRRequest) returns ( OCRResponse) {}
}

